### PR TITLE
[ADP-3272] WIP: Rough sketch of UTxO handling simplification.

### PR DESCRIPTION
**Pull request by jonathanknowles**
🏗️ This PR does not compile -- it's just a rough sketch of a potential simplification of `balanceTransaction`.

TODO:
- [ ] Specify the difference in the specification.
- [ ] PartialTxWithUTxO -- specify what will happen to this.
- [ ] 

## Issue

ADP-3272

### CODE REVIEW

File: lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs [hunk 1]
```diff
@@ -467,19 +465,29 @@ instance IsRecentEra era => Buildable (PartialTx era)
...
     }
 
 constructUTxOIndex
     :: IsRecentEra era
     => UTxO era
+    -- ^ Defines the complete set of UTxOs that may be referenced by a
+    -- transaction.
+    -> Set TxIn
+    -- ^ Defines the subset of UTxOs that may be spent in addition to any
+    -- UTxOs that are already spent by a transcation.
```
Comment (jonathanknowles) [inline lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Balance.hs:480]:
```suggestion
    -- UTxOs that are already spent by a transaction.
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 2]
```diff
@@ -3588,7 +3593,7 @@ balanceTransaction
 
         pure $ Write.PartialTx
             (Write.fromCardanoApiTx tx)
-            externalUTxO
+            mempty  -- externalUTxO
```
Comment (jonathanknowles) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:3596]: To be deleted ^^^

File: lib/wallet/src/Cardano/Wallet.hs [hunk 3]
```diff
@@ -2386,10 +2386,10 @@ buildTransactionPure
                 (Left $ unsafeShelleyOnlyGetRewardXPub @s (getState wallet))
                 txCtx
                 (Left preSelection)
+    let utxoAll = Write.fromWalletUTxO utxo
+    let utxoSpendable = Map.keysSet (unUTxO utxoAll)
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet.hs:2390]: In the case where there is no partial transaction available, we allow `balanceTx` to spend from the entire UTxO set.

**Review by Anviking — COMMENTED** [review]
While neat, I believe keeping the "extra utxo" alongside the `Tx` whose inputs it resolves makes conseptual sense. So I'd be hesitent in going away from this.

For instance, we could imagine writing the following helper function:
```haskell
addInput :: (TxIn, TxOut era) -> PartialTx era -> PartialTx era
addInput (inp, out) = -- add inp to tx, add (inp, out) to extraUTxO
```

I agree the wallet-influenced notion of UTxO conflict (different value or address) ideally shouldn't exist in `cardano-balance-tx`, but perhaps there is another solution to it which may become more apparent in the future. I don't think it's a problem worth thinking too much about at the moment.

**Comment by abailly** [comment]
Closing this as it's outdated and does not compile. Will revisit later.

**Pull request closed without merging**